### PR TITLE
Add system tests to generate scaffold

### PR DESCRIPTION
**Pull request by dinahshi**
`generate scaffold` currently generates a simple system test. This change
populates the generated system test with index, create, update, and destroy
behaviour. Feature request from @eileencodes 

Sample generated test from `bin/rails generate scaffold HighScore game:string score:integer`:
```
require "application_system_test_case"

class HighScoresTest < ApplicationSystemTestCase
  setup do
    @high_score = high_scores(:one)
  end

  test "visiting the index" do
    visit high_scores_url
    assert_selector "h1", text: "High Scores"
  end

  test "creating a High score" do
    visit high_scores_url
    click_on "New High Score"

    fill_in "Game", with: @high_score.game
    fill_in "Score", with: @high_score.score
    click_on "Create High score"

    assert_text "High score was successfully created"
    click_on "Back"
  end

  test "updating a High score" do
    visit high_scores_url
    click_on "Edit", match: :first

    fill_in "Game", with: @high_score.game
    fill_in "Score", with: @high_score.score
    click_on "Update High score"

    assert_text "High score was successfully updated"
    click_on "Back"
  end

  test "destroying a High score" do
    visit high_scores_url
    page.accept_confirm do
      click_on "Destroy", match: :first
    end

    assert_text "High score was successfully destroyed"
  end
end
```

### CODE REVIEW

File: railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb [hunk 1]
```diff
@@ -17,6 +17,10 @@ def create_test_files
         template_file = options.api? ? "api_functional_test.rb" : "functional_test.rb"
         template template_file,
                  File.join("test/controllers", controller_class_path, "#{controller_file_name}_controller_test.rb")
+
+        unless options.api?
+          template "system_test.rb", File.join("test/system", class_path, "#{file_name.pluralize}_test.rb")
+        end
```
Comment (eileencodes) [inline railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb]: I think this needs to check `depends_on_system_test?` because some users may want to skip system tests when generating the scaffold.

File: railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb [hunk 2]
```diff
@@ -30,17 +34,21 @@ def fixture_name
...
+        return {} if attributes_names.empty?
+
+        attributes_names.map do |name|
+          if %w(password password_confirmation).include?(name) && attributes.any?(&:password_digest?)
+            ["#{name}", "'secret'"]
+          else
+            ["#{name}", "@#{singular_table_name}.#{name}"]
+          end
+        end.sort.to_h
+      end
```
Comment (eileencodes) [inline railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb]: The rubocop build is failing for these because the methods should be indented 2 spaces under `private` and there should be a space after the bracket here `{|k, v| "#{k}: #{v}"}.join(", ")`.

File: railties/lib/rails/generators/test_unit/scaffold/templates/api_functional_test.rb [hunk 3]
```diff
@@ -29,7 +29,7 @@ class <%= controller_class_name %>ControllerTest < ActionDispatch::IntegrationTe
   end
 
   test "should update <%= singular_table_name %>" do
-    patch <%= show_helper %>, params: { <%= "#{singular_table_name}: { #{attributes_hash} }" %> }, as: :json
+    patch <%= show_helper %>, params: { <%= "#{singular_table_name}: { #{attributes_string} }" %> }, as: :json
```
Comment (eileencodes) [inline railties/lib/rails/generators/test_unit/scaffold/templates/api_functional_test.rb]: Why did these need to change?

**Review by eileencodes — CHANGES_REQUESTED** [review]
This is looking really great! I just tried it out and all the tests worked with the scaffold! 😄 

I left a few comments for things that need some changes. Thanks for working on this!

File: railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb (hunk 1 above)
Comment (dinahshi) [inline railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb]: Looks like this method is defined in [`Rails::Generators::AppBase`](https://github.com/rails/rails/blob/master/railties/lib/rails/generators/app_base.rb#L404-L406). Would it be best to replicate this in `TestUnit::Generators::Base` or rely on the `system_tests` option set [here](https://github.com/rails/rails/blob/master/railties/lib/rails/generators/rails/scaffold/scaffold_generator.rb#L19)?

File: railties/lib/rails/generators/test_unit/scaffold/templates/api_functional_test.rb (hunk 3 above)
Comment (dinahshi) [inline railties/lib/rails/generators/test_unit/scaffold/templates/api_functional_test.rb]: `attributes_hash` formerly referred to a string representation of the attributes (i.e. `"{key: value}"`). I needed a real ruby hash to iterate over so I named that method `attributes_hash` and renamed this to reflect its type.

Comment (eileencodes) [inline railties/lib/rails/generators/test_unit/scaffold/templates/api_functional_test.rb]: Cool thanks for the explanation!

File: railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb (hunk 1 above)
Comment (eileencodes) [inline railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb]: I think what you wrote is correct. I was thinking of the skip code elsewhere in the generators.

File: railties/lib/rails/generators/rails/scaffold/scaffold_generator.rb [hunk 4]
```diff
@@ -16,13 +16,10 @@ class ScaffoldGenerator < ResourceGenerator # :nodoc:
       def handle_skip
         @options = @options.merge(stylesheets: false) unless options[:assets]
         @options = @options.merge(stylesheet_engine: false) unless options[:stylesheets] && options[:scaffold_stylesheet]
-        @options = @options.merge(system_tests: false) if options[:api]
```
Comment (dinahshi) [inline railties/lib/rails/generators/rails/scaffold/scaffold_generator.rb]: I swapped this out for `@options = @options.merge(system_tests: nil) if options[:api]` and it still wasn't available in the options hash in railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb so I left the `options.api?` check in. I'm not sure what the purpose of this line is.

**Comment by eileencodes** [comment]
Awesome work @dinahshi! Can you squash your commits into 1 and then I'll merge? Thanks so much for working on this.

**Review by eileencodes — APPROVED** [review]

File: railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb [hunk 5]
```diff
@@ -11,12 +11,19 @@ class ScaffoldGenerator < Base # :nodoc:
...
       argument :attributes, type: :array, default: [], banner: "field:type field:type"
 
       def create_test_files
         template_file = options.api? ? "api_functional_test.rb" : "functional_test.rb"
         template template_file,
                  File.join("test/controllers", controller_class_path, "#{controller_file_name}_controller_test.rb")
+
+        unless options.api? || options[:system_tests].nil?
+          template "system_test.rb", File.join("test/system", class_path, "#{file_name.pluralize}_test.rb")
+        end
```
Comment (eileencodes) [inline railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb]: I just tested this and there's a missing test case. This doesn't work for generating a scaffold and passing the skip system test option. 

`bin/rails g scaffold Tree title:string --skip-system-test`

In that case the system test file would still be generated because we're not checking at all for the `skip-system-test` option which is why I ended up changing the config file in the other PR. Because then otherwise we end up checking for 2 options for skipping system tests.

**Review by eileencodes — CHANGES_REQUESTED** [review]
There's one issue when we're skipping system tests on the scaffold generator. There's a missing test case since all the tests are passing.

File: railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb (hunk 5 above)
Comment (dinahshi) [inline railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb]: I tried this on master and it's not currently supported. `bin/rails g scaffold Tree title:string --skip-test` also doesn't skip any tests (including model, controller, and system tests). As a result I think adding command line options to non-application generate commands is likely a larger task and should be separate from this change.

Comment (eileencodes) [inline railties/lib/rails/generators/test_unit/scaffold/scaffold_generator.rb]: You're right, it's not supported on master. I think this PR is fine then and if we want to implement `--skip-system-test` for scaffold we can do that in a separate PR.

**Review by eileencodes — APPROVED** [review]
